Condensed graph of reaction

hesther opened the pull request:
Allows (atom-mapped) reactions as input to Chemprop (keyword --reaction), which changes the molgraph (additional nodes + edges), as well as the atom and bond features. Furthermore, I included an option to have explicit hydrogens in the molecular graph (works for molecules and reactions as input), since some reaction datasets have mapped hydrogens, and some users might want to keep those. Let me know what you think!

hesther commented:
Seems like I made an error when copying over code - will fix this now

hesther commented:
Fixed - and ready to review (sorry for the last minute changes after pushing! I decided to put the arguments to training instead of common, and missed some changes associated with that). Should be good now.

hesther commented:
If you would like to test a minimal example:

`python train.py --data_path reactions.csv --dataset_type regression --reaction --explicit_h`

with reactions.csv (containing a few dummy reactions):

```
AAM,ea
[C:1]([c:2]1[n:3][o:4][n:5][n:6]1)([H:7])([H:8])[H:9]>>[C:1]1([H:7])([H:8])/[C:2](=[N:3]\[H:9])[N:6]1[N:5]=[O:4],77.6892865680395
[O:1]=[C:2]([N:3]=[N+:4]=[N-:5])[H:6]>>[N:3](=[N+:4]=[N-:5])[H:6].[O+:1]#[C-:2],53.326161667307
[O:1]([c+:2]1[n-:3][c:4]([H:8])[n:5][n:6]1[H:9])[H:7]>>[O:1](/[C:2]([N:3]1[C:4]([H:8])=[N:5]1)=[N:6]\[H:9])[H:7],79.71279521384959
[O:1]=[C:2]([C:3](=[O:4])[C:5]#[N:6])[H:7]>>[C:3](=[O:4])=[C:5]=[N:6][H:7].[O+:1]#[C-:2],31.5096944471283
[O:1]([c:2]1[n:3][o:4][c:5]([H:8])[c:6]1[H:9])[H:7]>>[O:1]([C:2](=[N-:3])[C:6](=[C+:5][O:4][H:8])[H:9])[H:7],83.8019575979882
[O:1]([C:2]([C:3]#[N:4])([C:5]#[N:6])[H:8])[H:7]>>[C:5]=[N:6][H:8].[O:1]([C+:2]=[C:3]=[N-:4])[H:7],71.0974346938255
[n:1]1([H:6])[c:2]([H:7])[c:3]([H:8])[n:4][n:5]1>>[N:1]1=[N:5][N:4]=[C:3]([H:8])[C:2]1([H:6])[H:7],50.9148303874494
[O:1]([c:2]1[c:3]([H:8])[n:4]([H:9])[n:5][c:6]1[H:10])[H:7]>>[O:1]([C@@:3]1([H:8])[C+:2]=[C:6]([H:10])[N-:5][N:4]1[H:9])[H:7],92.89835608525352
[O:1]=[C:2]([C:3]#[C:4][C:5]#[N:6])[H:7]>>[O-:1][N+:6]#[C:5][C:4]1=[C:3]=[C:2]1[H:7],143.977264352104
[O:1]=[C:2]([C:3]([C:4]#[C:5][H:9])([H:7])[H:8])[H:6]>>[O:1]([C:2](=[C:3]([H:7])[H:8])[H:6])[C:4]#[C:5][H:9],89.7649823417606
```

(Should yield about 29kcal/mol test RMSE with hydrogens and 23kcal/mol without hydrogens - let me know if you would like larger datasets to test)

You can also omit the `--explicit_h` flag, which will remove the hydrogens (although they are mapped)

cjmcgill commented on chemprop/features/featurization.py:
```diff
@@ -166,7 +218,19 @@ def __init__(self, mol: Union[str, Chem.Mol],
         # Convert SMILES to RDKit molecule if necessary
         if type(mol) == str:
-            mol = Chem.MolFromSmiles(mol)
+            if REACTION_MODE:
```
I might be mistaken, but how does this check work? Either reaction mode hasn't been set so this is None and it errors, or it's a string and passes. 

Should this be `if REACTION_MODE is not None`?

hesther replied:
If reaction mode hasn't been set, it is None (otherwise a string). The boolean value of None is False, and True for a string (or other objects that are not None), so the check works out fine - but I have to admit it is rather ugly ;) It is actually an artefact, since I started out with `REACTION_MODE` being Boolean. I will correct this to `if REACTION_MODE is not None` to make the code more readable.

mliu49 replied:
Relying on falsyness of `None` is generally not a problem. Using `is not None` is more specific and is better in situations where other falsy values are possible and need to pass the check (e.g. `0`, `''`).

hesther replied:
Since `REACTION_MODE` can only be `None`, `'Reac_Diff'`, `'Prod_Diff'` or `'Reac_Prod'`, I guess there is nothing that could currently go wrong  - but, to avoid bugs in future code, it might be easier to read/understand to just say `if REACTION_MODE != None`, so I am happy to change that!

mliu49 replied:
Sounds good. Also, the syntax in your previous comment was better; comparisons against `None` should be done using `is` rather than `=`.

cjmcgill replied:
I thought it was more unstable than that. In that case I have no problem here.

hesther replied:
Due to changes in the global variables REACTION and REACTION MODE, as well as a helper function to make RDKit molecules from strings, this should be better readable now

cjmcgill commented on chemprop/features/featurization.py:
```diff
@@ -178,57 +242,133 @@ def __init__(self, mol: Union[str, Chem.Mol],
-                    descr = bond_features_extra[bond.GetIdx()].tolist()
-                    if overwrite_default_bond_features:
-                        f_bond = descr
+        if not REACTION_MODE:
```
REACTION_MODE is not a Boolean is it?

hesther replied:
Same as above, the check works out, since `bool(None)=False` and `bool("anystring")=True` which is correct but ugly. Will correct this to `if REACTION_MODE == None`

hesther replied:
Due to the new treatment of the global variables, this is now nicer, too!

cjmcgill reviewed (commented):
I haven't tested it yet (and had trouble following in featurization) but did leave some code comments.

This is pretty complex. Do you want to add a pytest of this function as well?

hesther commented:
Thanks for all your comments, I will work through all of them. Yes, I think I should add some pytest functions, since this is a pretty large change, and as a high risk of breaking in future code changes if there are no proper tests. I will come up with some!

hesther commented on README.md:
```diff
@@ -242,6 +243,10 @@ The bond-level features are concatenated with the bond feature vectors before th
 
+### Reaction
+
+As an alternative to molecule SMILES, Chemprop can also process atom-mapped reaction SMILES, using the option `--reaction`, which transforms the inputted reaction to the corresponding condensed graph of reactions and changes the initial atom and bond features to hold information from both the reactant and product (option `--reaction_mode Reac_Prod`, default), or from the reactant and the differen ...
```
Will add a description. Reaction SMILES (actually, they are called SMIRKS) always look like: `reactant1.reactant2.reactant3>reagent1.reagent2>product1` or reactants>>product1.product2 if there are no reagents specified and multiple molecules are separated with dots. E.g. `CO >> C=O` means that methanol reacts to formaldehyde.

I will add an example to the text, as well as update "inputted" to "input" (it did sound a bit weird to me too).

This also solves some of the comments below, e.g. `CO => C=O`  would not be a valid reaction SMILES string per definition.

cjmcgill replied:
I did not realize that reaction smiles had a canonical format. Great!

hesther replied:
Updated readme to describe the format of reaction smiles, and a longer description of how reaction mode changes features.

hesther commented on chemprop/data/data.py:
```diff
@@ -123,9 +143,23 @@ def __init__(self,
-        mol = [SMILES_TO_MOL.get(s, Chem.MolFromSmiles(s)) for s in self.smiles]
+        if REACTION:
+            if EXPLICIT_H:
+                mol = [SMILES_TO_MOL.get(s, (Chem.MolFromSmiles(s.split(">")[0],sanitize=False),Chem.MolFromSmiles(s.split(">")[-1],sanitize=False))) for s in self.smiles]
```
See above, per definition of reaction smiles, e.g. in the Daylight manual [here](https://www.daylight.com/meetings/summerschool01/course/basics/smirks.html) the operator is always `>` (although there can be `>reagent>` or simply `>>`, which is why I split by `>` and use the first as reactants and the last/third as product (the second one would be reagents which are disregarded))

hesther commented on chemprop/data/data.py:
```diff
@@ -91,12 +104,19 @@ def __init__(self,
+                            self.features.extend(features_generator(m[0]))
+                        elif m[0] is not None and m[1] is not None and m[0].GetNumHeavyAtoms() == 0:
+                            self.features.extend(np.zeros(len(features_generator(Chem.MolFromSmiles('C')))))   
+                    
```
Good question, I have to look into this. I assume that with the get function you could access the variable from other scripts, too (which might be a better solution for EXPLICIT_H which is currently being set twice, as you correctly noted).

hesther replied:
I added get functions for EXPLICIT_H, REACTION and REACTION_MODE and moved all the global function getters and setters to chemprop/features/featurization.py. I can import them via the get functions if needed, for example in chemprop/data/data.py (which resolves the double EXPLICIT_H), and makes the code much cleaner!

hesther commented on chemprop/data/scaffold.py:
```diff
@@ -12,21 +12,26 @@
     """
-    mol = Chem.MolFromSmiles(mol) if type(mol) == str else mol
+    if type(mol) == tuple:
+        mol = mol[0]
```
That is a good question - not that I am aware of, but I think I should nevertheless support both options. Will change this.

hesther replied:
Changed to support all options (str of mol/reactions, mol of mol/reactions)

hesther commented on chemprop/data/scaffold.py:
```diff
@@ -12,21 +12,26 @@
 
 
-def generate_scaffold(mol: Union[str, Chem.Mol], include_chirality: bool = False) -> str:
+def generate_scaffold(mol: Union[str, Union[Chem.Mol,Tuple[Chem.Mol,Chem.Mol]]], include_chirality: bool = False) -> str:
```
Just googled - it does, will update throughout all scripts (this occurs multiple times)

hesther replied:
Updated to be a Union of multiple elements, also in other locations

hesther commented on chemprop/data/scaffold.py:
```diff
@@ -12,21 +12,26 @@
-    mol = Chem.MolFromSmiles(mol) if type(mol) == str else mol
+    if type(mol) == tuple:
+        mol = mol[0]
+    elif type(mol) == str and ">" in mol:
```
`mol` can either be 1) a string, molecule, 2) a RDKit molecule object, 3) a string, reaction, containing `>`, or 4) a tuple of RDKit molecules (and thus a reaction).
 
 The first if clause will only apply to case 4, and select the reactant molecules out of the tuple.
 
 The elif clause will detect case 3, and then create a RDKit molecule object. If `mol` is not a string (failure in first condition, `elif type(mol) == str`), the second condition (`">" in mol`) will not be executed/checked.
 
 The else clause (line 26-28) catches case 1 and 2.
 
 I just checked the logic of this again for a small test case, and it works fine for case 1-4.

cjmcgill replied:
Ah, this is a failure of my understanding of python. I assumed the second condition would be checked regardless.

hesther commented on chemprop/data/scaffold.py:
```diff
@@ -12,21 +12,26 @@
+    elif type(mol) == str and ">" in mol:
+        mol = Chem.MolFromSmiles(mol.split(">")[0])
+    else:
+        mol = Chem.MolFromSmiles(mol) if type(mol) == str else mol
```
To my understanding, in chemprop/data/scaffold.py, function `scaffold_split()` we check the number of molecules and raise a ValueError if it is not 1. Then, data.mols is flattened in line 79 (since it was already checked that number_of_molecules==1) (see [here](https://github.com/chemprop/chemprop/commit/2c98bf743cf90893dcc7b9f2b70786a2c7735b2f#diff-daff7c1aa3e0cd29613fa2f4be2ef1944f340b2da3e028a940c0f08dc62e33e7R79) for the corresponding commit, line 79 in chemprop/data/scaffold.py) - unfortunately, this is not very general, and I cannot use the same logic for splitting reactions.

cjmcgill replied:
Oh, I was not expecting it to be unsupported. Okay, this seems okay then. 

I feel a little odd about the explicit check for ">" in a string in the middle of something else. It seems like maybe there's an opportunity to have a utility function elsewhere that you can call to do that check (is_reaction_smiles() or something) and then have it available if you ever need that check elsewhere. This is fine though if that sounds unnecessary.

hesther commented on chemprop/features/featurization.py:
```diff
@@ -44,6 +45,22 @@ def get_atom_fdim(overwrite_default_atom: bool = False) -> int:
+def set_explicit_h_feat(explicit_h: bool) -> None:
+    r"""Sets whether RDKit molecules will be constructed with explicit Hs"""
+    global EXPLICIT_H
+    EXPLICIT_H = explicit_h
```
Yes, that is true - my experience with global variables is rather limited, I will research whether there is a better way to do this.

hesther replied:
I now condensed it down to a single EXPLICIT_H accessible from all modules

hesther commented on chemprop/train/cross_validate.py:
```diff
@@ -55,6 +56,15 @@ def cross_validate(args: TrainArgs,
     args.save(os.path.join(args.save_dir, 'args.json'))
 
+    #set explicit H option and reaction option
+    set_explicit_h_feat(args.explicit_h)
+    set_explicit_h(args.explicit_h)
```
I could condense them down!

hesther commented on chemprop/train/cross_validate.py:
```diff
@@ -55,6 +56,15 @@ def cross_validate(args: TrainArgs,
+    #set explicit H option and reaction option
+    set_explicit_h_feat(args.explicit_h)
+    set_explicit_h(args.explicit_h)
+    set_reaction(args.reaction)
+    if args.reaction:
+        set_reaction_mode(args.reaction_mode)
+        set_reaction_atom_fdim()
+        set_reaction_bond_fdim()
```
I could condense them down!

hesther replied:
Condensed all functions (5 previously) to 2, one to deal with explicit hydrogens, another to deal with setting reaction mode, feature dimensions, etc. This simplifies the code a great deal!

hesther commented on chemprop/train/make_predictions.py:
```diff
@@ -37,6 +38,16 @@ def make_predictions(args: PredictArgs, smiles: List[List[str]] = None) -> List[
     if args.bond_features_path is not None:
         set_extra_bond_fdim(train_args.bond_features_size)
 
+    #set explicit H option and reaction option
+    set_explicit_h(train_args.explicit_h)
+    set_explicit_h_feat(train_args.explicit_h)
+    set_reaction(train_args.reaction) 
+
+    if args.reaction:
+        set_reaction_mode(train_args.reaction_mode)
+        set_reaction_atom_fdim()
+        set_reaction_bond_fdim()
```
I could condense them down!

hesther commented on chemprop/train/molecule_fingerprint.py:
```diff
@@ -146,4 +157,4 @@ def chemprop_fingerprint() -> None:
     """
-    molecule_fingerprint(args=PredictArgs().parse_args())
\ No newline at end of file
+    molecule_fingerprint(args=PredictArgs().parse_args())
```
There was no "end-of-line" previously, which my text editor corrected automatically ;) I think it is good practice to have "end-of-lines" also on the last line, and other files of chemprop do.

cjmcgill replied:
Well gosh who left that hanging there with no line end (it was me)

hesther commented:
I have worked through all of the current comments, and I think found a good way of treating the global variables in a consistent way (they are now all in one single module, and can be accessed via get and set functions from other modules). Also, the functions to set reaction mode have been condensed down to one, and I introduced a new helper function to make molecules from strings (either as reaction or molecule, or keeping hydrogens or not). Since I could not find any utility script that would not cause circular imports by using it in `chemprop/data/data.py` and `chemprop/features/featurization.py`, I introduced a new one (`chemprop/rdkit.py`).

I also resolved a merge conflict which was introduced when fixing the typing of the MPNN forward pass yesterday.

Let me know if there are any other changes I could make, or if you see any conflicts which I might have overlooked. I will furthermore think about some tests soon and add them to the `tests/test_integration.py`, but maybe you could already review the rest of the code?

hesther commented:
I have now added four tests to `tests/test_integration.py` (reaction mode, reaction mode + scaffold split, reaction mode + Morgan fingerprint features, reaction mode + explicit_h). Since graphs and molecule objects are cached for all tests (that is, within each python script), I furthermore introduced a command line option which forces the cache to be cleared if set to True, so I can recalculate atom/bond features between tests (e.g. testing with explicit_h after testing without explicit_h).

mliu49 reviewed (commented):
Sorry for the delay in reviewing this. I added a bunch of minor code style comments, but I also have two general points to raise for discussion:
1. I'm not a big fan of toggling functionality via the `is_reaction` attribute. I think it would be better to take an object-oriented approach and create (for example) a `ReactionGraph` class analogous to the `MolGraph` class. This could help improve code readability and maintainability. However, there are a bunch of classes involved (`MoleculeDataset`, `MoleculeDatapoint`, etc.) and they aren't currently designed to be subclassed, so it may be complicated to refactor.
2. Using global variables is generally best avoided if possible, especially if they're used across multiple modules. This is mainly because they make code hard to understand and can behave oddly if you're not familiar with how they work in Python. I think the need for global variables is also somewhat a result of the first point, so switching to an object-oriented approach could also help eliminate them.

Unfortunately, I don't have really concrete suggestions at the moment for how to best address these two points, mainly because I'm not sufficiently familiar with how all of these parts interact. I'd be happy to discuss this more offline to come up with a reasonable approach (considering time and effort).

hesther commented on chemprop/args.py:
```diff
@@ -321,6 +328,21 @@ class TrainArgs(CommonArgs):
+    """
+    Whether to adjust MPNN layer to take reactions as input instead of molecules.
+    """
+    reaction_mode: Literal['Reac_Prod','Reac_Diff','Prod_Diff'] = 'Reac_Diff'
```
I totally agree, have changed it!

hesther commented on chemprop/data/data.py:
```diff
@@ -28,9 +29,15 @@ def set_cache_graph(cache_graph: bool) -> None:
 
 
+def empty_cache():
+    r"""Empties the cache of :class:`~chemprop.features.MolGraph`\ s  and RDKit molecules."""
```
Not really, changed!

hesther commented on chemprop/data/data.py:
```diff
@@ -123,10 +140,10 @@ def __init__(self,
         """Gets the corresponding list of RDKit molecules for the corresponding SMILES list."""
-        mol = [SMILES_TO_MOL.get(s, Chem.MolFromSmiles(s)) for s in self.smiles]
-
+        mol = make_mols(self.smiles,self.is_reaction,self.is_explicit_h)
```
Added commas here (and throughout other lines and files).

hesther commented on chemprop/data/scaffold.py:
```diff
@@ -9,24 +9,26 @@
-    mol = Chem.MolFromSmiles(mol) if type(mol) == str else mol
+    if type(mol) == str:
+        mol = make_mol(mol,keep_h=False)
+    if type(mol) == tuple:
```
Nice, I didn't now that! Have changed it throughout different files.

hesther commented on chemprop/data/scaffold.py:
```diff
@@ -9,24 +9,26 @@
 
-
-def generate_scaffold(mol: Union[str, Chem.Mol], include_chirality: bool = False) -> str:
+def generate_scaffold(mol: Union[str, Chem.Mol, Tuple[Chem.Mol,Chem.Mol]], include_chirality: bool = False) -> str:
```
Thanks, I didn't know that, and have changed it where appropriate.

hesther commented on chemprop/data/utils.py:
```diff
@@ -143,7 +143,8 @@ def filter_invalid_smiles(data: MoleculeDataset) -> MoleculeDataset:
                             if all(s != '' for s in datapoint.smiles) and all(m is not None for m in datapoint.mol)
-                            and all(m.GetNumHeavyAtoms() > 0 for m in datapoint.mol)])
+                            and all(m.GetNumHeavyAtoms() > 0 for m in datapoint.mol if type(m)!=tuple)
+                            and all(m[0].GetNumHeavyAtoms() + m[1].GetNumHeavyAtoms() > 0 for m in datapoint.mol if type(m)==tuple)])
```
Done!

hesther commented on chemprop/features/featurization.py:
```diff
@@ -134,6 +177,35 @@ def bond_features(bond: Chem.rdchem.Bond) -> List[Union[bool, int, float]]:
+    :param mol_prod: An RDKit molecule of the products.
+    :return: A dictionary of corresponding reactant and product atom indices.
+    """
+    only_prod_ids=[]
```
Changed here and elsewhere throughout different lines and files. Sorry that I didn't do that in the first place!

hesther commented:
Thanks for reviewing the code! I have taken into account all the style changes in the most recent commit.

Regarding ReactionGraph objects: I totally agree, having a ReactionGraph object would be much nicer - the reason I didn't do that in the first place is because all corresponding functions and classes acting on MolGraph would need to be adapted then, and I don't quite know how to do that without again introducing either a global variable, or passing the arguments object to all of them. I will message you offline - I am happy to change that, but don't quite know where and how to start.

mliu49 reviewed (approved):
To summarize offline discussion, it would require significant refactoring of many core classes to implement an object-oriented approach to supporting reaction input. So this implementation will be merged as-is, and refactoring can be done later if reaction support becomes a development priority.
